controller/volume_controller.go

```diff
@@ -1125,77 +1182,413 @@ func (vc *VolumeController) ReconcileVolumeState(v *longhorn.Volume, es map[stri
+				if err := vc.openVolumeDependentResources(v, e, rs, log); err != nil {
+					return err
+				}
+				if vc.verifyVolumeDependentResourcesOpened(e, rs) {
+					v.Status.CurrentNodeID = v.Spec.NodeID
+					v.Status.State = longhorn.VolumeStateAttached
+					vc.eventRecorder.Eventf(v, v1.EventTypeNormal, constant.EventReasonAttached, "volume %v has been attached to %v", v.Name, v.Status.CurrentNodeID)
+				}
+			}
+			return nil
+		}
+
+		if v.Status.CurrentNodeID != "" {
+			if v.Spec.NodeID == v.Status.CurrentNodeID {
+				switch v.Status.State {
+				case longhorn.VolumeStateAttaching, longhorn.VolumeStateDetached:
+					vc.closeVolumeDependentResources(v, e, rs)
+					v.Status.State = longhorn.VolumeStateDetaching
+				case longhorn.VolumeStateDetaching:
+					vc.closeVolumeDependentResources(v, e, rs)
+					if vc.verifyVolumeDependentResourcesClosed(e, rs) {
+						v.Status.CurrentNodeID = ""
+						v.Status.State = longhorn.VolumeStateDetached
+						vc.eventRecorder.Eventf(v, v1.EventTypeNormal, constant.EventReasonDetached, "volume %v has been detached", v.Name)
+					}
+				case longhorn.VolumeStateAttached:
+					// This is a stable state
+					// Try to openVolumeDependentResources so that we start the newly added replicas if they exist
+					if err := vc.openVolumeDependentResources(v, e, rs, log); err != nil {
+						return err
+					}
+					if !vc.verifyVolumeDependentResourcesOpened(e, rs) {
+						log.Warnf("volume is attached but dependent resources are not opened")
+					}
+				}
+				return nil
+			}
+			if v.Spec.NodeID != v.Status.CurrentNodeID {
+				switch v.Status.State {
+				case longhorn.VolumeStateDetached, longhorn.VolumeStateAttaching:
+					vc.closeVolumeDependentResources(v, e, rs)
+					v.Status.State = longhorn.VolumeStateDetaching
+				case longhorn.VolumeStateDetaching:
+					vc.closeVolumeDependentResources(v, e, rs)
+					if vc.verifyVolumeDependentResourcesClosed(e, rs) {
+						v.Status.CurrentNodeID = ""
+						v.Status.State = longhorn.VolumeStateDetached
+						vc.eventRecorder.Eventf(v, v1.EventTypeNormal, constant.EventReasonDetached, "volume %v has been detached", v.Name)
+					}
+				case longhorn.VolumeStateAttached:
+					if v.Spec.Migratable && v.Spec.AccessMode == longhorn.AccessModeReadWriteMany && v.Status.CurrentMigrationNodeID != "" {
+						if err := vc.openVolumeDependentResources(v, e, rs, log); err != nil {
+							return err
+						}
+						if !vc.verifyVolumeDependentResourcesOpened(e, rs) {
+							log.Warnf("volume is attached but dependent resources are not opened")
+						}
+					} else {
+						vc.closeVolumeDependentResources(v, e, rs)
+						v.Status.State = longhorn.VolumeStateDetaching
+					}
+				}
+				return nil
+			}
+		}
+	}
+
+	return nil
+}
+
+func (vc *VolumeController) reconcileVolumeCreation(v *longhorn.Volume, e *longhorn.Engine, es map[string]*longhorn.Engine, rs map[string]*longhorn.Replica) (bool, *longhorn.Engine, error) {
```

Comment:
Why return `*longhorn.engine`

Reply:
So the main loop in volume controller get the newly created engine:
```
isNewVolume, e, err := vc.reconcileVolumeCreation(v, e, es, rs)
```